executor/slow_query.go

```diff
@@ -34,56 +37,57 @@ import (
-	{execdetails.RequestCountStr, mysql.TypeLonglong, 20, mysql.UnsignedFlag, nil, nil},
-	{execdetails.TotalKeysStr, mysql.TypeLonglong, 20, mysql.UnsignedFlag, nil, nil},
-	{execdetails.ProcessKeysStr, mysql.TypeLonglong, 20, mysql.UnsignedFlag, nil, nil},
-	{variable.SlowLogDBStr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogIndexNamesStr, mysql.TypeVarchar, 100, 0, nil, nil},
-	{variable.SlowLogIsInternalStr, mysql.TypeTiny, 1, 0, nil, nil},
-	{variable.SlowLogDigestStr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogStatsInfoStr, mysql.TypeVarchar, 512, 0, nil, nil},
-	{variable.SlowLogCopProcAvg, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopProcP90, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopProcMax, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopProcAddr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogCopWaitAvg, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopWaitP90, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopWaitMax, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopWaitAddr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogMemMax, mysql.TypeLonglong, 20, 0, nil, nil},
-	{variable.SlowLogSucc, mysql.TypeTiny, 1, 0, nil, nil},
-	{variable.SlowLogPlan, mysql.TypeLongBlob, types.UnspecifiedLength, 0, nil, nil},
-	{variable.SlowLogPlanDigest, mysql.TypeVarchar, 128, 0, nil, nil},
-	{variable.SlowLogPrevStmt, mysql.TypeLongBlob, types.UnspecifiedLength, 0, nil, nil},
-	{variable.SlowLogQuerySQLStr, mysql.TypeLongBlob, types.UnspecifiedLength, 0, nil, nil},
+//SlowQueryRetriever uses to read slow log data.
+type SlowQueryRetriever struct {
+	table       *model.TableInfo
+	outputCols  []*model.ColumnInfo
+	retrieved   bool
+	initialized bool
+	file        *os.File
+	fileLine    int
 }
 
-func dataForSlowLog(ctx sessionctx.Context) ([][]types.Datum, error) {
+func (e *SlowQueryRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.retrieved {
+		return nil, nil
+	}
+	if !e.initialized {
+		var err error
+		e.file, err = os.Open(sctx.GetSessionVars().SlowQueryFile)
+		if err != nil {
+			return nil, err
+		}
+		e.initialized = true
+	}
+	fullRows, err := e.dataForSlowLog(sctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(e.outputCols) == len(e.table.Columns) {
+		return fullRows, nil
+	}
+	rows := make([][]types.Datum, len(fullRows))
+	for i, fullRow := range fullRows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		rows[i] = row
+	}
+	return rows, nil
+}
+
+func (e *SlowQueryRetriever) close() error {
+	if e.file != nil {
+		err := e.file.Close()
+		if err != nil {
+			logutil.BgLogger().Error("close slow log file failed.", zap.Error(err))
+		}
+	}
+	return nil
+}
+
+func (e *SlowQueryRetriever) dataForSlowLog(ctx sessionctx.Context) ([][]types.Datum, error) {
```

Comment:
How about remove the e.file and e.fileLine, both of them can be get  from `ctx sessionctx.Context`, which is a parameter for `dataForSlowLog()`, and change this method to a func seems much better.

Reply:
Ok, It seems both of them would be cached in e to reduce cost.